### DISCUSSION TITLE
mult keyboards with same layout implemenations ideas

### DISCUSSION THREAD

**Discussion by molleweide** [kmonad/kmonad#390] [Ideas]
Yo!!!

I have thought about it and it would be nice if you could create a tiny defsrc file for each keyboard you own and then 
manage the actual custom layout you use with kmonad in a single file. I start a discussion thread on this here.

For example, in my use case, I use a 3 by 5 minimal layout which means that my layout fits within any keyboard that exists out there, so therefore I am imagining something like this:

``` bash

# defsrc'es
- macbook_pro_intel_defsrc.kbd
- m1_mac.kbd
- my_wierd_qmk_keyboard.kbd

# main kmonad layout
- main.kbd

```

Then, the next step is to modify the parser or something in the core that accounts for this.
I am imagining that one sources the `main.kbd` into each defsrc and you also would have to
specify which positions in the keyboard array that would correspond to the main layout.

I am not so good at haskell still but if I get good enough I want to implement this.
It feels like you'd take the current parser functionality and make it into a two step process kind of.

This idea makes sense, right?? :D

**Comment by MuhammedZakir** [comment]
IIRC, something like this was discussed previously in an issue. I think this we will be easy to do when the whole code refactor is done (especially the work in `tinker` branch).

Btw, you can use a shell script to do what you want.

```zsh
#!/sbin/env zsh

local defsrc="${K_DEFSRC:-$HOME/.config/kmonad/defsrc.kbd}"
local baselayer="${K_BASELAYER:-$HOME/.config/kmonad/ansi_base.kbd}"

# $1 => config
# $2-$n => kmonad args
exec kmonad =(cat "$defsrc" "$baselayer" "$1") "$@[2,-1]"
```

E.g.

```sh
env K_BASELAYER=$HOME/.config/kmonad/iso_base.kbd $HOME/.local/bin/kmonad_m $HOME/.config/kmonad/80keys.kbd
````

**Reply by molleweide** [reply]
woaw, I have to try this!!

**Reply by tydazzz1** [reply]
env K_BASELAYER=$HOME/.config/kmonad/iso_base.kbd $HOME/.local/bin/kmonad_m $HOME/.config/kmonad/80keys.kbd

**Comment by molleweide** [comment]
@MuhammedZakir if you have time may I ask you for your opinion on this because I am unsure if this will work or not.

I have been looking for a while now at your code snippet above  and re-read the tutorial.kbd, and my question is how I specify where in the defsrc that my sublayout should be mapped to. As an example I have provided my own layouts that I want to use
below. First comes my three hardware keyboards into which I want to map my minimal layout which is located at the bottom.
I have replaced all chars with XXs and underscores to make it easy to see how things should fit together.
Will this work, and if so would you please mind explaining to me how I can map the minimal layout into the three above.
I am not sure if I am asking for too much here but hopefully it will be helpful for others as well.

I also realize above that the laptop keyboards have 5 bottom row keys and the ergodox has 6, therefor I am mapping
both inner thumbs to `space` in order to have the same muscle memory everywhere . So in the ideal case I'd like to
ignore one of the inner-thumb-keys on the laptop layouts.

```clojure
;; 1. Below, are my three hardware keyboard layouts.
;; I have replaced the keys that are supposed to be replaced
;; by my 5 by 3 layout with underscores to make it easy to see
;; what I am trying to achieve.

(deflayer BASE_ERGODOX
  XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX
  XX   _    _    _    _    _    XX   XX   _    _    _    _    _    XX
  XX   _    _    _    _    _              _    _    _    _    _    XX
  XX   _    _    _    _    _    XX   XX   _    _    _    _    _    XX
  XX   XX   _    _    _                        _    _    _    XX   XX
                           XX   XX   XX   XX
                                XX   XX
                      XX   XX   XX   XX   XX   XX
)

;; in the layouts below, which are my standard lapo keyboards, I have shifted
;; `zxcvb` one step to the left in order to achieve a more comfortable 3 by 5
;; layout

(deflayer BASE_MACBOOK_PRO_SWE
  XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX
       XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX
  XX   _    _    _    _    _         _    _    _    _    _    XX   XX   XX
  XX   _    _    _    _    _         _    _    _    _    _    XX   XX
  XX   _    _    _    _    _    XX   _    _    _    _    _    XX   XX
  XX   XX   _    _              _              _    _    XX   XX   XX
)

(deflayer BASE_MACBOOK_AIR
  XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX
  XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX
  XX   _    _    _    _    _         _    _    _    _    _    XX   XX   XX
  XX   _    _    _    _    _         _    _    _    _    _    XX   XX
       _    _    _    _    _    XX   _    _    _    _    _    XX   XX
  XX   XX   _    _              _              _    _    XX   XX   XX
)

;; 2. And now follows my minimal layout. that should go into the layouts
;; above. 

(deflayer MINIMAL_SUBLAYOUT
    XX    XX    XX    XX    XX         XX    XX    XX    XX    XX
    XX    XX    XX    XX    XX         XX    XX    XX    XX    XX
    XX    XX    XX    XX    XX         XX    XX    XX    XX    XX
                XX    XX    XX         XX    XX    XX
)
```

**Reply by MuhammedZakir** [reply]
~For any of these keyboards, do you want to remap keys that are not in minimal layer?~

Nevermind. Understood from this:
```
;; 2. And now follows my minimal layout. that should go into the layouts
;; above. 
```

**Reply by molleweide** [reply]
I am sort of thinking that each defsrc file would look something like this below. I copy pasted from my research file so there are random aliases in the layout positions below. There is a deflayer in each defsrc file that XXs out all the surrounding keys so that only the minimal layout passes through.

```clojure
(defcfg
  input (iokit-name "Apple Internal Keyboard / Trackpad")
  output (kext)
  fallthrough true
)

(defsrc
  esc  f1   f2   f3   f4   f5   f6   f7   f8   f9   f10  f11  f12
  grv  1    2    3    4    5    6    7    8    9    0    -    =    bspc
  tab  q    w    e    r    t    y    u    i    o    p    [    ]    \
  caps a    s    d    f    g    h    j    k    l    ;    '    ret
       lsft z    x    c    v    b    n    m    ,    .    /    rsft up
  fn   lctl lalt lmet           spc            rmet ralt left down rght
)

(deflayer 
  _    XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX
  XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX   XX
  XX   @b00 @b01 @b02 @b03 @b04      @b05 @b06 @b07 @b08 @b09 XX   XX   XX
  XX   @b10 @b11 @b12 @b13 @b14      @b15 @b16 @b17 @b18 @b19 XX   XX
       @b20 @b21 @b22 @b23 @b24 XX   @b25 @b26 @b27 @b28 @b29 XX   XX
  XX   XX   esc  @tl2           @spc1          @tr2 esc  XX   XX   XX
)
```

**Reply by MuhammedZakir** [reply]
It is allowed. But using `XX` will disable those buttons. Isn't fallthrough (`_`) better for this?

I am confused now. Does this mean that you don't want to use buttons that are not in the "sub-layer"?

**Reply by molleweide** [reply]
I'm sorry I didn't mean to confuse you. I don't want to use they keys that are outside of the minimal layout. I want to use the minimal layout ONLY.  That is why I am/want to surround the minimal layout with XXs.

**Reply by MuhammedZakir** [reply]
If that's the case, then things are simpler. Just use the minimal layout in everything, including `defsrc` and base layer. And in `defcfg`, don't set `fallthrough` to `true` (by default, it is `false`k). Disabling `fallthrough` will make KMonad ignore all keys that are not in the `defsrc`, essentially blocking them.

Because KMonad has cli commands to change `defcfg`, you only need to have one config, and then use shell script(s) to change `input` and `output`.

1. In `~/.config/kmonad/common-layout.kbd`:

```clojure
(defcfg
  input (iokit-name "common layout") ;; just a placeholder
  output (kext)
)

(defsrc
  ;; your common layout
)

(deflayer baselayer
  ;; your common layout
)

;; etc...
```

2. In `~/.config/kmonad/ergodox.sh`:

```zsh
#!/sbin/env zsh

local -A keyboards=( [ergodox.sh]="<ergodox's product name"
                     [mac-air.sh]="<product name>"
                     [mac-pro.sh]="<product name>"
                   )

local input="(iokit-name \"${keyboards[$0]}\")"

kmonad common-layout.kbd --fallthrough false --input "$input"
```

Note: don't forget to replace each `<product name>` with corresponding product name.

3. In `~/.config/kmonad/`:

```zsh
ln ergodox.sh mac-air.sh
ln ergodox.sh mac-pro.sh
```

---

This is one way to do it. Another way is to use a single shell script and decide the product name based on argument to the script.

**Reply by molleweide** [reply]
Thank you very much for your detailed reply.  There is one thing that I don't understand that I still need to ask you that doesn't make sense to me: 

How do I know that the minimal layout will be in the right place on each of my keyboards? 

I do understand everything in the above when it comes to how you are linking each of the files together but how the minimal layout is placed in the correct position is a mystery to me. I mean, don't I have to do some kind of thing where I
tell kmonad how I want to map the minimal layout into each of the keyboards? I am assuming here that I put the minimal layout into the 'common-layout' above but how will my macbook keyboard know how it is positionned?

However, it is really cool to see how you did the above.

**Reply by molleweide** [reply]
I am going to try to follow your idea above and then see if I can do it in a single script as you say on bottom.

**Reply by MuhammedZakir** [reply]
`defsrc` defines what keys you want to remap and defines the initial layout. You can think of it like this

user presses a key > KMonad checks that key in `defsrc` > if it exists, get the position of that key in `defsrc` > execute button in that position in the topmost layer

It is NOT

user presses a key > KMonad gets that key's (physical) position in keyboard > execute the button in that position in topmost layer

---

We use the keyboard layout as the representation in `defsrc`/`deflayer` to make things easier _for us_. To KMonad, the order or the structure doesn't matter. You can put everything in a single line or one key per line, or whatever. It doesn't matter, as long as `deflayer`s matches the `defsrc`.

**Reply by molleweide** [reply]
Aaaaahhaaaa so I could even put the defsrc in the reverse order if I wanted since it looks for what tokens I want to replace. Now I finally understand.

**Reply by molleweide** [reply]
Hey Muhammed!

I did it by catting everything in a function with a loop. What do you think of this?

```zsh
compile_layout() {
    echo "$defcfg"
    cat $defsrc
    for f in "$config_kmonad_dir/common/"*.kbd; do
        cat "$f"
    done
}
compile_layout
kmonad =(compile_layout)
```

## question

do you know what the  equivalent of `=(...)` is for `bash`. I have been playing around and I only get this to work
with you suggested `=(...)` from above.

## full script

I have put `defsrc` files in one dir and then I keep all of the common modules under `common`. I needed to do this
because it is a different `defsrc` for each of my keyboards. Next step is to build this script out further and see what happens.

<details>

```zsh
#!/usr/bin/env zsh

script_path=$(realpath $0)
config_kmonad_dir=$(dirname $script_path)
# config_kmonad_dir2="${BASH_SOURCE%/*}"

if test -z ${1-}; then
    echo "No input specified. You need to specify what key board"
    echo "you want to run kmonad for."
    echo "(ergo|mbair|mbpro)"
    exit 1
fi

# --------------------------
# ---       cfg       ---
# --------------------------

declare -A defcfg_input_str=(
    [ergo]="Ergodox EZ"
    [mbair]="Apple Internal Keyboard / Trackpad"
    [mbpro]="Apple Internal Keyboard / Trackpad"
)

input="${defcfg_input_str[$1]}"
output="kext"
fallthrough="false"

defcfg="(defcfg
input (iokit-name \"$input\")
output ($output)
fallthrough $fallthrough
)"

# --------------------------
# ---       src       ---
# --------------------------

declare -A defsrc_files=(
    [ergo]="defsrc_ergodox.kbd"
    [mbair]="defsrc_mbair.kbd"
    [mbpro]="defsrc_mbpro_swe.kbd"
)

defsrc="$config_kmonad_dir/defsrc/${defsrc_files[$1]}"

# ----------------------------------
# ---      compile and run      ---
# ----------------------------------

compile_layout() {
    echo "$defcfg"
    cat $defsrc
    for f in "$config_kmonad_dir/common/"*.kbd; do
        cat "$f"
    done
}
kmonad =(compile_layout) --dry-run
```

</details>

**Reply by MuhammedZakir** [reply]
> I did it by catting everything in a function with a loop. What do you think of this?

You don't need a loop: `cat "$config_kmonad_dir/common/"*.kbd`.

> do you know what the equivalent of =(...) is for bash. I have been playing around and I only get this to work
with you suggested =(...) from above.

I don't know as I am not familiar with Bash. `=(<cmd>)` creates a temporary file with the command's output and replaces `=(<md>)` with that filename. If Bash doesn't have it, you can manually do that.
```sh
local config_final="$(mktemp /tmp/kmonad-XXX)"
kmonad "$config_final"
```

> [...] I needed to do this because it is a different defsrc for each of my keyboards. [...]

In macOS, things that maybe different in `defsrc` are `input` and `fallthrough`. Why not set them using cli flags?

**Reply by molleweide** [reply]
1. Thanks for pointing out that I don't need loop for cat
2. I found that `<()` is the equivalent for bash.
3. Do I use the flag `--defsrc` then? It does not show when kmonad is listing flags?

Can I write it something like `kmonad <(common) --defcfg $cfg --defsrc $src

I was thinking that both input and output can be variable if I am jumping between OSes so that is why I did it in a string.
Oh but I guess I can do all of these in a flag also I realize. So the main question is if the `--defsrc` flag works?

**Reply by molleweide** [reply]
oh wait now I see I missunderstood, you meant the components within the defcfg are possible to set via flags. Never mind my questions above

**Reply by molleweide** [reply]
It seems that fallthrough sais that `Invalid argument 'false'` if I pass it via flag?!

**Reply by molleweide** [reply]
but it works when I pass it in the defcfg string

**Reply by molleweide** [reply]
also it turns out that the bash command `<()` didn't really work it seems. This is the output:
```
kmonad: /dev/fd/63: openFile: invalid argument (Bad file descriptor)
```
However, the zsh `=()` worked perfectly and I can run kmonad which is awesome.

I am going to test some other things and see if something else works with bash.

**Reply by molleweide** [reply]
Okay so now I got it to work thanks to your suggestion of making a temp file. This is the result in bash:

```bash
compile_layout() {
    echo "$defcfg"
    cat $defsrc
    cat "$config_kmonad_dir/common/"*.kbd
}

# bash
tmp_file="$(mktemp /tmp/kmonad-XXX)"
compile_layout > $tmp_file
sudo kmonad $tmp_file
# sudo kmonad <(compile_layout) # doesn't work
# -f doesn't accept `false`. Complains about invalid...
# kmonad <(compile_layout) -f $fallthrough -i $input -o $output -d

# zsh
# sudo kmonad =(compile_layout)
```

**Reply by MuhammedZakir** [reply]
`<()` is different from `=()`. You can check it using `echo <() , =()` in ZSH. To know the difference, see `PROCESS SUBSTITUTION` in `zshexpn` manpage and (no `=()`) `Process Substitution` in `bash` manpage.

```man
<(list):
       If  the system supports the /dev/fd mechanism, the command argument is the name of the device file
       corresponding to a file descriptor; otherwise, if the system supports named pipes  (FIFOs),
       the  command  argument  will  be  a  named  pipe.

=(list):
       If =(...) is used instead of <(...), then the file passed as an argument will be the name of a temporary file  con‐
       taining the output of the list process.  This may be used instead of the < form for a program that expects to lseek
       (see lseek(2)) on the input file.
```

---

About `fallthrough` and `allow-cmd`: looks like `-ctrue` and `-ftrue` works? By default, these are set to `false`, so you only need to use these flags if you want to enable them. Btw, I only tested whether `kmonad` succesfully parses these cli args, didn't test whether `fallthrough` and `allow-cmd` works.

**Reply by molleweide** [reply]
Awesome! I am going to read up on these and get back later. Process substitution is a good keyword. It is hard to search for these things when one doesn't know these konds of key words.

**Reply by MuhammedZakir** [reply]
I usually search using symbols rather than keywords. In this case, `<(`. Searching using a partial sequence usually get better results.

**Reply by MuhammedZakir** [reply]
Which version of KMonad are you using?

**Reply by molleweide** [reply]
nice yeah I have started to understand that searching like that is good. which search engine to do you mostly search for code when you learn?

kmo version = `0.4.1, commit 1e432897dd5b7f8b2c2202e577ef22c58fdc5e5a`

**Reply by MuhammedZakir** [reply]
> nice yeah I have started to understand that searching like that is good. which search engine to do you mostly search for code when you learn?

I use DuckDuckGo and Startpage.

> kmo version = `0.4.1, commit 1e432897dd5b7f8b2c2202e577ef22c58fdc5e5a`

`master` branch or `keycode-refactor` branch?

**Reply by molleweide** [reply]
Master branch

**Comment by manna-harbour** [comment]
Here's a work around using a macro processor: #435.

**Reply by molleweide** [reply]
Cool i'll check it out!! I think I saw this project or similar before when I got my first qmk keyboard. It is interesting to shift the keyb up and use mult thumb keys!

**Reply by molleweide** [reply]
> Here's a work around using a macro processor: #435.

Does miryoku implement mouse keys? And if so, to what extent?

**Reply by manna-harbour** [reply]
The mapping with the shifted rows does work, and you can use it, but it is very weird. You might use it if you've completely forgotten how to use a normal keyboard and you suddenly had no choice.  But there are a couple of other options included. The"lite" mapping keeps the traditional home positions and adds only alphas, home row mods, and the Nav layer, so that should be completely usable. But I think the main use would be on non-programmable ergo boards, or if you're considering a programmable ergo board but want to try the keymap out first. 

On other Miryoku implementations I used macro to map the keymap onto different physical layouts, but in KMonad that's not needed as `defsrc` does the same thing.

Miryoku KMonad uses [mouse keys on the host](https://github.com/manna-harbour/miryoku_kmonad#mouse-keys) for the mouse layer, but I might add some alternatives from #370.

**Reply by molleweide** [reply]
Cool, i am using a 5x3 layout already. I will see if I can try this layout with my setup.

Are the host mouse keys on macos hard coded to the chars as per the docs or can one reprogram the keys? Ideally I am learning haskell so that I can add some good mouse keys myself.

**Reply by manna-harbour** [reply]
You can [use](https://github.com/manna-harbour/miryoku_kmonad#use-with-other-keymaps) the Miryoku KMonad makefile directly with any keymap, so you could use it to add conditional `defsrc` to your existing keymap.  And as your keymap is 3x5+3, you could even recreate it with Miryoku Babel and apply it to the other Miryoku implementations, then you can use it on any supported keyboard and firmware.

I've only used the X11 mouse keys.  I believe the other OSs use the same bindings, but I don't know if any support changing them.  The functionality is a bit limited though, so one of the other mouse keys implementations might be better.
